File: source/kCura.Relativity.Export/FileObjects/ExportFileSerializer.vb
Imports System.Xml.Linq
Imports kCura.Relativity.Export.Types
Imports Relativity

Namespace kCura.Relativity.Export.FileObjects
	Public Class ExportFileSerializer
		Private _settingsValidator As New ExportSettingsValidator
		Public Property SettingsValidator As ExportSettingsValidator
			Get
				If _settingsValidator Is Nothing Then _settingsValidator = New ExportSettingsValidator
				Return _settingsValidator
			End Get
			Set(ByVal value As ExportSettingsValidator)
				_settingsValidator = value
			End Set
		End Property

		Public Overridable Function TransformExportFileXml(ByVal input As XDocument) As String
			Return input.ToString
		End Function

		Public Overridable Function DeserializeExportFile(ByVal currentExportFile As ExportFile, ByVal xml As String) As ExportFile
			Dim retval As New ExportFile(currentExportFile.ArtifactTypeID)
			Dim deserialized As ExportFile = Me.DeserializeExportFile(XDocument.Parse(xml))
			For Each p As System.Reflection.PropertyInfo In (From prop As System.Reflection.PropertyInfo In retval.GetType.GetProperties Where prop.CanWrite)
				p.SetValue(retval, p.GetValue(If(PropertyIsReadFromExisting(p), currentExportFile, deserialized), Nothing), Nothing)
			Next
			'TODO: test
			Select Case retval.TypeOfExport
				Case ExportFile.ExportType.AncestorSearch, ExportFile.ExportType.ParentSearch
					retval.ArtifactID = currentExportFile.ArtifactID
				Case ExportFile.ExportType.Production
					retval.ImagePrecedence = New Pair() {}
			End Select
			If Not SqlNameHelper.GetSqlFriendlyName(currentExportFile.ObjectTypeName).Equals(SqlNameHelper.GetSqlFriendlyName(retval.ObjectTypeName)) Then
				retval = New ErrorExportFile("Cannot load '" & currentExportFile.ObjectTypeName & "' settings from a saved '" & retval.ObjectTypeName & "' export")
			End If
			If Not Me.SettingsValidator.IsValidExportDirectory(retval.FolderPath) Then retval.FolderPath = String.Empty
			Return retval
		End Function

		Private Function PropertyIsReadFromExisting(ByVal p As System.Reflection.PropertyInfo) As Boolean
			For Each att As Attribute In p.GetCustomAttributes(GetType(ReadFromExisting), False)
				Return True
			Next
			Return False
		End Function

		Public Overridable Function DeserializeExportFile(ByVal xml As XDocument) As ExportFile
			Dim deserializer As New System.Runtime.Serialization.Formatters.Soap.SoapFormatter
			Dim cleansedInput As String = Me.TransformExportFileXml(xml)
			Dim sr As New System.IO.MemoryStream(System.Text.Encoding.UTF8.GetBytes(cleansedInput))
			Dim deserialized As ExportFile = Nothing
			Try
				deserialized = DirectCast(deserializer.Deserialize(sr), ExportFile)
			Catch
				Throw
			Finally
				sr.Close()
			End Try
			Return deserialized
		End Function

		Public Class ExportSettingsValidator
			Public Overridable Function IsValidExportDirectory(ByVal path As String) As Boolean
				Return System.IO.Directory.Exists(path)
			End Function
		End Class
	End Class
End Namespace

File: source/kCura.Relativity.Export.ConvertedToC#/FileObjects/ExportFileSerializer.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Xml.Linq;
using kCura.Relativity.Export.Types;
using Relativity;

namespace kCura.Relativity.Export.FileObjects
{
	public class ExportFileSerializer
	{
		private ExportSettingsValidator _settingsValidator = new ExportSettingsValidator();
		public ExportSettingsValidator SettingsValidator {
			get {
				if (_settingsValidator == null)
					_settingsValidator = new ExportSettingsValidator();
				return _settingsValidator;
			}
			set { _settingsValidator = value; }
		}

		public virtual string TransformExportFileXml(XDocument input)
		{
			return input.ToString();
		}

		public virtual ExportFile DeserializeExportFile(ExportFile currentExportFile, string xml)
		{
			ExportFile retval = new ExportFile(currentExportFile.ArtifactTypeID);
			ExportFile deserialized = this.DeserializeExportFile(XDocument.Parse(xml));
			foreach (System.Reflection.PropertyInfo p in retval.GetType().GetProperties().Where(prop => prop.CanWrite)) {
				p.SetValue(retval, p.GetValue(PropertyIsReadFromExisting(p) ? currentExportFile : deserialized, null), null);
			}
			//TODO: test
			switch (retval.TypeOfExport) {
				case ExportFile.ExportType.AncestorSearch:
				case ExportFile.ExportType.ParentSearch:
					retval.ArtifactID = currentExportFile.ArtifactID;
					break;
				case ExportFile.ExportType.Production:
					retval.ImagePrecedence = new List<Pair>();
					break;
			}
			if (!SqlNameHelper.GetSqlFriendlyName(currentExportFile.ObjectTypeName).Equals(SqlNameHelper.GetSqlFriendlyName(retval.ObjectTypeName))) {
				retval = new ErrorExportFile("Cannot load '" + currentExportFile.ObjectTypeName + "' settings from a saved '" + retval.ObjectTypeName + "' export");
			}
			if (!this.SettingsValidator.IsValidExportDirectory(retval.FolderPath))
				retval.FolderPath = string.Empty;
			return retval;
		}

		private bool PropertyIsReadFromExisting(System.Reflection.PropertyInfo p)
		{
			foreach (Attribute att in p.GetCustomAttributes(typeof(ReadFromExisting), false)) {
				return true;
			}
			return false;
		}

		public virtual ExportFile DeserializeExportFile(XDocument xml)
		{
			System.Runtime.Serialization.Formatters.Soap.SoapFormatter deserializer = new System.Runtime.Serialization.Formatters.Soap.SoapFormatter();
			string cleansedInput = this.TransformExportFileXml(xml);
			System.IO.MemoryStream sr = new System.IO.MemoryStream(System.Text.Encoding.UTF8.GetBytes(cleansedInput));
			ExportFile deserialized = null;
			try {
				deserialized = (ExportFile)deserializer.Deserialize(sr);
			} catch {
				throw;
			} finally {
				sr.Close();
			}
			return deserialized;
		}

		public class ExportSettingsValidator
		{
			public virtual bool IsValidExportDirectory(string path)
			{
				return System.IO.Directory.Exists(path);
			}
		}
	}
}

